### ISSUE TITLE
-50% hashrate on single core with more than 2 threads (Big issue)

### ISSUE THREAD

**Issue by unudoitreipatru** [fireice-uk/xmr-stak#1966]
All was fine in previous version, now in v8 I can use only 2 threads on a phisical core, more than that is dropping hashrate dramaticly. I've checked everything, nothing seems to work, is the miner.

**Comment by Spudz76** [comment]
Old telemetry was inaccurate, overdriving CPUs used to make it *show* a higher rate but that was just due to starving the telemetry of CPU power to measure hashrate properly.

If you watch your pool side effective hashrate it will not be different.  Old version was lying about true hashrate (or, estimating it wrong maybe not lying).

I ran LPM=5 on some cpus because it *showed* +20H/s but once I gave up on that being correct and only relied on what the pool showed, everything is the same effective rate.

There should be no gain from overblowing your cache, that forces some threads off onto system RAM which is slow compared to cache.  The apparent gains in old versions were fake.

**Comment by unudoitreipatru** [comment]
I have to not agree to you, I've check myself hashrate on pool side and was 15-20% increase from 4th to 5th on a single core. Maybe you cannot see much difference on one core but i run more than 100, and belive me, was better. Now is gone 50% but I have hope some dev will see into that, is a big loss on Xeons

**Comment by unudoitreipatru** [comment]
Check pool side hashrate, is what I've said

**Comment by Spudz76** [comment]
@SChernykh see if this users setup is truly correct and maybe find out why it is not performant

**Comment by SChernykh** [comment]
Intel Atom without AES? 50% of previous hashrate is your new normal, sorry.

**Comment by SChernykh** [comment]
@unudoitreipatru What's your CPU model? Can you paste your config file?

**Comment by unudoitreipatru** [comment]
Example E5-2650 v3 on ubuntu 16 
2 threads, 1 core - 88H/s
3 threads, 1 core - 62H/s
5 threads, 1 core - 56H/s
"cpu_threads_conf" :
[
    { "low_power_mode" : 2, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 0 },

],

On previous version was over 150 H/s

**Comment by unudoitreipatru** [comment]
Sorry, cannot use more than 1 core, server is working for other jobs, but was fine with v7

**Comment by SChernykh** [comment]
@unudoitreipatru then it won't work fast. One physical core can't do many parallel threads, new algorithm is heavy on computations.

**Comment by unudoitreipatru** [comment]
You have said on previous topic that you'll try to optimize for 4 threads on 1 physical core, this will be one big help for those using spare cores. Many thanks

**Comment by SChernykh** [comment]
Yes, this is in my plans, but I haven't made good enough version yet.

**Comment by Spudz76** [comment]
But my Atom does 6KH I'm super happy.

That was just showing that telemetry can and does lie when threads are overloaded.  But I also believe the pool somewhat, I was only showing that I wasn't wrong about telemetry possibly being invalid locally in some configurations (or other side-loads which it sounds like you have on this server).

I did not lose any rate on pool-side despite showing a 20H loss going LPM=1/false versus LPM=5 and it lying to me.

**Comment by StingeyNinja** [comment]
I just wanted to chime in and report that I'm observing exactly the same issue as @unudoitreipatru.  I run with 5 threads on a single core, and since the changeover to v8 I'm seeing approx. half the hash rate being reported both in xmr-stack directly (pressing 'h') and on the pool-side. 

### CPU

**Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz**

Maximum speed:	2.29 GHz
Sockets:	1
Virtual processors:	1
Virtual machine:	Yes

### cpu.txt
`"cpu_threads_conf" :
[
    { "low_power_mode" : 5, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 0 },

],`

**Comment by foomor** [comment]
Same here, 40% down on Xeon E5-2670 v2 (25MB L3 cache)

v7:
2 cores, x5: **320**H/s
2 cores, x2: 201H/s
2 cores, x1: 117H/s

v8:
2 cores, x5: 130H/s
2 cores, x2: **190**H/s
2 cores, x1: 113H/s

**Comment by Spudz76** [comment]
```
vendor_id       : GenuineIntel
cpu family      : 6
model           : 45
model name      : Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
stepping        : 7
microcode       : 0x710
cpu MHz         : 3021.687
cache size      : 20480 KB
```

I comment out the top two and bottom two cores otherwise default.
```
//    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 0 },
//    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 1 },
    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 2 },
    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 3 },
    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 4 },
    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 5 },
    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 6 },
    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 7 },
//    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 16 },
//    { "low_power_mode" : false, "no_prefetch" : true, "asm" : "auto", "affine_to_cpu" : 17 },
```

About 50H/s per active core, dual server, gets 600H/s still right now, verified CN2v2 according to my proxy (xmr-stak set to automatic `monero` not forced generic `cryptonight_v8`).  But, my entire miner-proxy-pool chain is not showing any problems and did the fork on their own.

**Comment by unudoitreipatru** [comment]
I see your point, you don't use multithreads on any core, this wasn't subject in this topic. Issue appear on 3 or more threads on one physical core, you have "low_power_mode" : false" , problem is when settings like this : "low_power_mode" : 3" or more.

**Comment by bilias** [comment]
> On this I saw zero difference in speed across the fork:
>
> About 50H/s per active core, dual server, gets 600H/s still right now, verified CN2v2 according to my proxy >(xmr-stak set to automatic `monero` not forced generic `cryptonight_v8`). But, my entire miner-proxy-pool >chain is not showing any problems and did the fork on their own.

You didn't have optimal configuration on v7 that's why you don't see big difference now.
With 4 cores (instead of 6) you could get ~107/core with "low_power_mode" : 3 ~= 430/cpu instead of 300

As I see it in v7 we could utilize more CPU cache to get better hashes.
In v8 as it seems so far, asm optimization is up to x2 cache. Putting more cache does not help.
So the impact is on users utilizing few cores but x[3-5] cache.

Also for you configuration now you can still get better results:
4 cores/ "low_power_mode" : 2 -> 354/cpu
6 cores/ "low_power_mode" : 2 -> 460/cpu

**Comment by Spudz76** [comment]
I tested nearly every combination and this was the best perf I ever saw.  Any form of LPM>1 got worse always, or the hashrate would wander around the threads like it was in contention/waitstates.

Interesting.

Machine also runs a bunch of "real day job" VMs thus my leaving some resources unmined (so the VMs don't "notice") and maybe why I never saw any gain from LPM besides 1.

**Comment by Spudz76** [comment]
Nope, I get 200 per 6cores with lpm2 vs constant 300 with single, just retested.  Both versions of algo same speeds.
Gotta be my sideload getting in the way of more / breaking the gain from stacking.  Similar to the new sqrt+div maneuver must break whatever your lucky harmonic was (SmartCache not being so Smart?)
